events: Markdown improvements

zecakeh opened the pull request:
- Improve detection of markdown.
- Enable strikethrough and tables.

Closes #1297.
















<!-- Replace -->
----
Preview: https://pr-1343--ruma-docs.surge.sh
<!-- Replace -->

jplatte commented on crates/ruma-common/src/events/message.rs:
```diff
@@ -238,12 +238,31 @@ impl Text {
+                pulldown_cmark::Event::Start(tag)
+                    | pulldown_cmark::Event::End(tag)
+                    if matches!(tag, pulldown_cmark::Tag::Paragraph)
+            )
```
Maybe introduce some bindings here to make it a little more readable, i.e.

```rust
let is_foo = matches!(event, …);
let is_bar = matches!(event, …);

!is_foo && !is_bar
```

zecakeh committed:
events: Fix markdown content detection

Encoded HTML entities would be detected as a change.
Use the parsed events instead to check if markdown syntax was detected.

zecakeh committed:
events: Enable more markdown syntax

Enable markdown extensions that produces spec-compliant HTML.

zecakeh commented:
Sorry, I changed almost everything so the push force makes the history clean again.

jplatte commented on crates/ruma-common/tests/events/room_message.rs:
```diff
@@ -302,6 +302,24 @@ fn markdown_content_serialization() {
+    formatted_body.unwrap();
+
+    // HTML entities don't trigger markdown.
+    let formatted_body = FormattedBody::markdown("A message with & HTML < entities");
```
This is an interesting case. When will HTML start having an effect? Is `<span>Hello</span>` considered a plain-text message?

zecakeh replied:
No, as soon as there is something that looks like a HTML tag it will be detected. At first I had written this test as `"A message with & HTML <entities>"` and it detected `<entities>` as HTML.

jplatte replied:
Can you add a test for that?

zecakeh replied:
Done
